File: apps/driver-android/src/main/kotlin/lk/mageride/driver/profile/ProfileEditors.kt
package lk.mageride.driver.profile

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Contacts
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import lk.mageride.driver.R
import lk.mageride.driver.onboarding.PhoneNumber
import lk.mageride.driver.onboarding.endonym
import lk.mageride.driver.onboarding.englishName
import lk.mageride.driver.ui.component.LabelledTextField
import lk.mageride.driver.ui.component.MageRideCta
import lk.mageride.driver.ui.component.PhoneNumberField
import lk.mageride.driver.ui.component.SelectionBox
import lk.mageride.driver.ui.theme.ControlTokens
import lk.mageride.driver.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.Language

/**
 * SCR-DA-029's three editors, each a modal bottom sheet.
 *
 * A sheet rather than a destination for each: none of the three is drawn as a screen in the
 * wireframe — the ✎ on the header row and the ✎ on the emergency-contact row are affordances *on*
 * the profile, and AL-35's ruling on SCR-DA-030's rating sheet ("a modal bottom sheet, not an
 * inline card") is the same shape applied to the same kind of edit.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun ProfileEditorSheet(
    state: DriverProfileState,
    viewModel: DriverProfileViewModel,
    modifier: Modifier = Modifier,
) {
    val sheet = state.sheet ?: return

    ModalBottomSheet(
        onDismissRequest = viewModel::dismissSheet,
        modifier = modifier,
        sheetState = rememberModalBottomSheetState(),
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(
                    start = MageRideTheme.spacing.md,
                    end = MageRideTheme.spacing.md,
                    bottom = MageRideTheme.spacing.lg,
                ),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
        ) {
            when (sheet) {
                ProfileSheet.Name -> NameEditor(state = state, viewModel = viewModel)
                ProfileSheet.Emergency -> EmergencyContactEditor(state = state, viewModel = viewModel)
                ProfileSheet.Language -> LanguageEditor(state = state, viewModel = viewModel)
            }
        }
    }
}

/** The header ✎ — `PUT /v1/users/me` with a display name (US-1.5). */
@Composable
private fun NameEditor(state: DriverProfileState, viewModel: DriverProfileViewModel) {
    Text(
        text = stringResource(R.string.profile_name_title),
        style = MaterialTheme.typography.titleMedium,
        color = MaterialTheme.colorScheme.onSurface,
    )
    LabelledTextField(
        label = stringResource(R.string.profile_name_label),
        value = state.nameDraft,
        onValueChange = viewModel::onNameChange,
    )
    MageRideCta(
        label = stringResource(R.string.action_save),
        onClick = viewModel::saveName,
        enabled = state.canSaveName,
        loading = state.saving,
    )
}

/**
 * AL-13's emergency contact — the name and number **SCR-DA-032's driver SOS sends to** (D-33).
 *
 * The picker is offered above the fields rather than instead of them: an address book is the fast
 * path and typing is the one that always works, and safety-svc answers
 * `400 no-emergency-contact` to an SOS raised by an account with none.
 */
@Composable
private fun EmergencyContactEditor(state: DriverProfileState, viewModel: DriverProfileViewModel) {
    val pickContact = rememberContactPicker { name, phone ->
        viewModel.onContactPicked(name = name, phone = phone)
    }

    Text(
        text = stringResource(R.string.profile_emergency_title),
        style = MaterialTheme.typography.titleMedium,
        color = MaterialTheme.colorScheme.onSurface,
    )
    Text(
        text = stringResource(R.string.profile_emergency_why),
        style = MaterialTheme.typography.bodySmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
    TextButton(onClick = pickContact) {
        Icon(
            imageVector = Icons.Outlined.Contacts,
            contentDescription = null,
            modifier = Modifier
                .size(ControlTokens.RowIcon)
                .padding(end = MageRideTheme.spacing.xxs),
        )
        Text(text = stringResource(R.string.profile_emergency_pick))
    }
    LabelledTextField(
        label = stringResource(R.string.profile_emergency_name_label),
        value = state.contactNameDraft,
        onValueChange = viewModel::onContactNameChange,
    )
    PhoneNumberField(
        value = state.contactPhoneDraft,
        onValueChange = viewModel::onContactPhoneChange,
        isError = state.contactPhoneRejected,
        supporting = stringResource(R.string.profile_emergency_phone_invalid)
            .takeIf { state.contactPhoneRejected },
        placeholder = PhoneNumber.PLACEHOLDER,
        prefix = PhoneNumber.COUNTRY_CODE,
    )
    MageRideCta(
        label = stringResource(R.string.action_save),
        onClick = viewModel::saveEmergencyContact,
        enabled = state.canSaveContact,
        loading = state.saving,
    )
}

/** AL-26's three boxes again — the same control SCR-DA-002 uses, and the same endonym table. */
@Composable
private fun LanguageEditor(state: DriverProfileState, viewModel: DriverProfileViewModel) {
    Text(
        text = stringResource(R.string.profile_language_title),
        style = MaterialTheme.typography.titleMedium,
        color = MaterialTheme.colorScheme.onSurface,
    )
    Language.entries.forEach { language ->
        SelectionBox(
            label = language.endonym,
            selected = state.profile?.language == language,
            onSelect = { viewModel.chooseLanguage(language) },
            secondary = language.englishName,
        )
    }
}

File: apps/driver-ios/DriverApp/Profile/ProfileEditors.swift
import MageRideShared
import SwiftUI

/// SCR-DI-029's three editors, each a `.sheet` at the `.medium` detent.
///
/// A sheet rather than a destination for each: none of the three is drawn as a screen in the wireframe
/// — the ✎ on the header row and the ✎ on the emergency-contact row are affordances *on* the profile —
/// and AL-35's ruling on SCR-DI-030's rating sheet (*"a modal bottom sheet, not an inline card"*) is
/// the same shape applied to the same kind of edit. `.medium` for the same reason SCR-DI-033a takes it:
/// the profile stays visible behind, which is what says the edit is a detour rather than a screen.
@MainActor
struct ProfileEditorSheet: View {

    let sheet: ProfileSheet
    @ObservedObject var model: DriverProfileModel

    @Environment(\.dismiss) private var dismiss

    var body: some View {
        NavigationStack {
            Group {
                switch sheet {
                case .name: NameEditor(model: model)
                case .emergency: EmergencyContactEditor(model: model)
                case .language: LanguageEditor(model: model)
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
            .background(MageRideColor.background)
            .navigationTitle(Text(key: titleKey))
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button { dismiss() } label: { Text(key: "action_cancel") }
                }
            }
        }
        .presentationDetents([.medium, .large])
    }

    private var titleKey: String {
        switch sheet {
        case .name: return "profile_name_title"
        case .emergency: return "profile_emergency_title"
        case .language: return "profile_language_title"
        }
    }
}

/// The header ✎ — `PUT /v1/users/me` with a display name (US-1.5).
private struct NameEditor: View {

    @ObservedObject var model: DriverProfileModel

    var body: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
            LabelledTextField(
                labelKey: "profile_name_label",
                value: Binding(get: { model.state.nameDraft }, set: model.onNameChange)
            )

            Button {
                Task { await model.saveName() }
            } label: {
                Text(key: "action_save")
            }
            .buttonStyle(.mageCta(loading: model.state.isSaving))
            .disabled(!model.state.canSaveName)

            if let errorKey = model.state.errorKey {
                FormErrorText(messageKey: errorKey)
            }
        }
        .padding(MageRideSpacing.md)
    }
}

/// AL-13's emergency contact — the name and number **SCR-DI-032's driver SOS sends to** (D-33).
///
/// The picker is offered above the fields rather than instead of them: an address book is the fast path
/// and typing is the one that always works, and safety-svc answers `400 no-emergency-contact` to an SOS
/// raised by an account with none.
private struct EmergencyContactEditor: View {

    @ObservedObject var model: DriverProfileModel

    @State private var isPickingContact = false

    var body: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
            Text(key: "profile_emergency_why")
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)

            Button { isPickingContact = true } label: {
                HStack(spacing: MageRideSpacing.xxs) {
                    Image(systemName: "person.crop.circle.badge.plus")
                        .font(.footnote)
                    Text(key: "profile_emergency_pick")
                        .mageFont(.bodyEmphasis)
                }
                .foregroundStyle(MageRideColor.primary)
            }
            .buttonStyle(.plain)

            LabelledTextField(
                labelKey: "profile_emergency_name_label",
                value: Binding(get: { model.state.contactNameDraft }, set: model.onContactNameChange)
            )

            VStack(alignment: .leading, spacing: MageRideSpacing.xxs) {
                SectionLabel(key: "profile_emergency_phone_label")
                PhoneNumberField(
                    value: Binding(get: { model.state.contactPhoneDraft }, set: model.onContactPhoneChange),
                    isError: model.state.isContactPhoneRejected
                )
                if model.state.isContactPhoneRejected {
                    FormErrorText(messageKey: "profile_emergency_phone_invalid")
                }
            }

            Button {
                Task { await model.saveEmergencyContact() }
            } label: {
                Text(key: "action_save")
            }
            .buttonStyle(.mageCta(loading: model.state.isSaving))
            .disabled(!model.state.canSaveContact)

            if let errorKey = model.state.errorKey {
                FormErrorText(messageKey: errorKey)
            }
        }
        .padding(MageRideSpacing.md)
        .sheet(isPresented: $isPickingContact) {
            ContactPickerView(
                onPicked: { name, phone in
                    isPickingContact = false
                    model.onContactPicked(name: name, phone: phone)
                },
                onDismiss: { isPickingContact = false }
            )
            .ignoresSafeArea()
        }
    }
}

/// AL-26's three boxes again — the same control SCR-DI-002 uses, and the same endonym table.
private struct LanguageEditor: View {

    @ObservedObject var model: DriverProfileModel

    var body: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.xs) {
            // `LanguageCityState.languages` rather than a second list: AL-26 fixes the order — Sinhala
            // first and default — and SCR-DI-002 is where it is written down.
            GroupedList {
                ForEach(Array(LanguageCityState.languages.enumerated()), id: \.offset) { index, language in
                    SelectionRow(
                        label: LanguageDisplay.endonym(language),
                        secondary: LanguageDisplay.englishName(language),
                        isSelected: model.state.profile?.language == language,
                        showsSeparator: index < LanguageCityState.languages.count - 1,
                        onSelect: { Task { await model.choose(language: language) } }
                    )
                }
            }

            if let errorKey = model.state.errorKey {
                FormErrorText(messageKey: errorKey)
            }
        }
        .padding(MageRideSpacing.md)
    }
}
